pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -97,84 +82,114 @@ func NewDRAPluginClient(pluginName string) (Client, error) {
 	}
 
 	return &draPluginClient{
-		pluginName:          pluginName,
-		addr:                draAddr(existingPlugin.endpoint),
-		nodeV1ClientCreator: newV1NodeClient,
+		pluginName:        pluginName,
+		addr:              draAddr(existingPlugin.endpoint),
+		nodeClientCreator: newNodeClient,
 	}, nil
 }
 
-func (r *draPluginClient) NodePrepareResource(
+func (r *draPluginClient) NodePrepareResources(
 	ctx context.Context,
-	namespace string,
-	claimUID types.UID,
-	claimName string,
-	resourceHandle string,
-) (*drapbv1.NodePrepareResourceResponse, error) {
-	klog.V(4).InfoS(
-		log("calling NodePrepareResource rpc"),
-		"namespace", namespace,
-		"claimUID", claimUID,
-		"claimName", claimName,
-		"resourceHandle", resourceHandle)
-
-	if r.nodeV1ClientCreator == nil {
-		return nil, errors.New("failed to call NodePrepareResource. nodeV1ClientCreator is nil")
+	req *drapb.NodePrepareResourcesRequest,
+	opts ...grpc.CallOption,
+) (resp *drapb.NodePrepareResourcesResponse, err error) {
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info(log("calling NodePrepareResources rpc"), "request", req)
+	defer logger.V(4).Info(log("done calling NodePrepareResources rpc"), "response", resp, "err", err)
+
+	if r.nodeClientCreator == nil {
+		return nil, errors.New("failed to call NodePrepareResources. nodeClientCreator is nil")
 	}
 
-	nodeClient, closer, err := r.nodeV1ClientCreator(r.addr)
+	nodeClient, nodeClientOld, closer, err := r.nodeClientCreator(r.addr)
 	if err != nil {
 		return nil, err
 	}
 	defer closer.Close()
 
-	req := &drapbv1.NodePrepareResourceRequest{
-		Namespace:      namespace,
-		ClaimUid:       string(claimUID),
-		ClaimName:      claimName,
-		ResourceHandle: resourceHandle,
-	}
-
 	ctx, cancel := context.WithTimeout(ctx, PluginClientTimeout)
 	defer cancel()
 
-	return nodeClient.NodePrepareResource(ctx, req)
+	resp, err = nodeClient.NodePrepareResources(ctx, req)
+	if err != nil {
+		status, _ := grpcstatus.FromError(err)
+		if status.Code() == grpccodes.Unimplemented {
```

Comment:
Is there no way to just check which API a plugin registered with, rather than attempt a call on the new API can fall back to the old one if it fails? If not, is there some way to make this more generic instead of inline in this call?

Reply:
OK, let's leave this as is and I'll work with @TommyStarK  to clean this up in his PR.